src/wt.ts

```diff
@@ -0,0 +1,357 @@
+  }
+  return token;
+}
+
+const retrieveFromQueryString = (search: string) => {
+  const qs = QS.parse(search, { ignoreQueryPrefix: true });
+  return qs.wvt as string | undefined;
+};
+
+const retrieveVisitorToken = (
+  config: Cookie.CookieAttributes = {},
+  search: string
+) =>
+  retrieveFromQueryString(search) ||
+  retrieveFromCookie(VISITOR_TOKEN_KEY, config);
+
+const retrievePageUUIDToken = () => retrieveFromCookie(PAGE_UUID_KEY);
+
+type WTConfig = {
+  cookies?: Cookies.CookieAttributes;
+  trackerUrl?: string;
+  trackerDomain?: string;
+  stringifyOptions?: QS.IStringifyOptions;
+  debounce?: {
+    min?: number;
+    max?: number;
+  };
+};
+
+/** @internal */
+export type WTContext = {
+  location: {
+    hostname: string;
+    href: string;
+    search: string;
+  };
+  document: {
+    referrer: string;
+  };
+  navigator: {
+    userAgent: string;
+  };
+  innerWidth: number;
+  innerHeight: number;
+  Image: typeof window.Image;
+};
+
+type WTEvent = {
+  kind?: string;
+  category?: string;
+  action?: string;
+  label?: string;
+  value?: string | number | boolean | null;
+  page_name?: string;
+  container?: string;
+  position?: string | number;
+  object_type?: string;
+  object_name?: string;
+  metadata?: Record<string, any>;
+  url: string;
+  referrer: string | undefined;
+  page_uuid: string | null;
+  ts: number;
+};
+
+export type WTEventParams = {
+  kind?: string;
+  category?: string;
+  action?: string;
+  label?: string;
+  value?: string | number | boolean | null;
+  pageName?: string;
+  container?: string;
+  position?: string | number;
+  objectType?: string;
+  objectName?: string;
+  [metadataKey: string]: any;
+};
+
+export type WTPayload = {
+  events: WTEvent[];
+  dimensions: {
+    width: number;
+    height: number;
+  };
+  agent: string;
+  rts: number;
+  wvt: string | undefined;
+};
+
+export class WT {
+  public loading = false;
+
+  private emitter = new EventEmitter();
+  private wtConfig: WTConfig = {
+    cookies: { expires: EXPIRES_IN_DAYS },
+  };
+  private paramDefaults = {};
+  private eventQueue: WTEvent[] = [];
+  private pageUuid: string | null = null;
+  private context: WTContext;
+  private loaderImage: HTMLImageElement | undefined;
+  private processEventsDebounced!: ReturnType<typeof debounce>;
+
+  constructor(context: WTContext) {
+    this.context = context;
+    this.updateProcessEventsDebounced();
+  }
+
+  public initialize(payload: WTConfig | ((config: WTConfig) => WTConfig)) {
+    this.config(payload);
+    if (this.wtConfig.cookies) {
+      this.getUUIDToken();
+      this.getVisitorToken();
+    }
+  }
+
+  public getVisitorToken() {
+    return retrieveVisitorToken(
+      this.wtConfig.cookies,
+      this.context.location && this.context.location.search
+    );
+  }
+
+  public getUUIDToken() {
+    this.pageUuid = retrievePageUUIDToken();
+    return this.pageUuid;
+  }
+
+  public track(kind: string, payload: WTEventParams = {}) {
```

Comment:
can we make `kind` be a type since we shouldn't allow arbitrary kinds?

Reply:
i'm a bit hesitant to constrain this. although we currently have a very well defined list (`event | pageview | pageload`), i'm on the fence of whether we should enforce that on a type level. i can think of a few options:

1. constrain directly in this signature, if a consumer wanted to override they could cast something to `any`
2. add a generic to `WT` that defaults to the current options (would probably go ahead and refactor the weird `withContext` factory pattern and export the class directly in this case)
3. leave as is

(my vote is 2 or 3)

Reply:
kind of like the idea of #2 just to reduce programmer error, but we can think more about this later in the spirit of moving this along.

Reply:
also, can we not use `payload` as the arg name? especially confusing since we have `WTPayload`